CONTRIBUTING.md

```diff
@@ -1,16 +1,721 @@
+
+```hcl
+output "project_id" {
+  description = "Project id."
+  value       = "${local.prefix}${var.name}"
+  depends_on = [
+    google_project.project,
+    data.google_project.project,
+    google_project_organization_policy.boolean,
+    google_project_organization_policy.list,
+    google_project_service.project_services,
+    google_compute_shared_vpc_service_project.service_projects,
+    google_project_iam_member.shared_vpc_host_robots,
+    google_kms_crypto_key_iam_member.service_identity_cmek,
+    google_project_service_identity.servicenetworking,
+    google_project_iam_member.servicenetworking
+  ]
+}
+```
+
+#### Why we don't use random strings in names
+
+This is more a convention than a design principle, but it's still important enough to be mentioned here: we **never use random resources for naming** and instead rely on an optional `prefix` variable which is implemented in most modules.
+
+This matches actual use where naming is a key requirement that needs to integrate with company-wide CMDBs and naming schemes used on-prem or in other clouds, and usually is formed by concatenating progressively more specific tokens (something like `myco-gcp-dev-net-hub-0`).
+
+Our approach supports easy implementation of company-specific policies and good readability, while still allowing a fairly compact way of ensuring unique resources have unique names.
+
+```hcl
+# prefix = "foo-gcp-dev"
+
+module "project" {
+  source = "./modules/project"
+  name   = "net-host-0"
+  prefix = var.prefix
+}
+
+module "project" {
+  source = "./modules/project"
+  name   = "net-svc-0"
+  prefix = var.prefix
+}
+```
+
+### FAST stage design
+
+Due to their increased complexity and larger scope, FAST stages have some additional design considerations. Please refer to the [FAST documentation](./fast/) for additional context.
+
+#### Standalone usage
+
+Each FAST stage should be designed so that it can optionally be used in isolation, with no dependencies on anything other than its variables.
+
+#### Stage interfaces
+
+Stages are designeded based on the concept of ["contracts" or interfaces](./fast/README.md#contracts-and-stages), which define what information is produced by one stage via outputs, which is then consumed by a following stage via variables.
+
+Interfaces are compact in size (few variables) but broad in scope (variables typically leverage maps), so that consumers can declare in variable types only the bits of information they are interested in.
+
+Resource management stages for example only export three map variables: `folder_ids`, `service_accounts`, `tag_names`. Those variables contain values for all the relevant created resources, but consumers are only interested in some of them and only need to declare those: networking stages for example only declare the folder and service account names they need.
```

Comment:
s/Resource management stages for example/For example, resource management stages/
s/created resources/resources created/

I'd remove the last part about networking, it's confusing since the paragraph already starts referring to the resman stages.

Reply:
it's actually the piece of code that demonstrates in practice how this works: multiple keys produced, only a subset declared and consumed